Playwright E2E tests: Multiple signed in roles

tarhi-saad opened the pull request:
This PR moves the following tests to playwright and removes the puppeteer equivalent:

- Guest user can place order
- Logged in user can place an order

**Setting up Playwright's multiple sign up roles**

We are using [Multiple signed in roles](https://playwright.dev/docs/auth#multiple-signed-in-roles) to test for authenticated vs. unauthenticated users. There are some required adjustments to fix https://github.com/woocommerce/woocommerce-blocks/pull/10561#issuecomment-1691870998. We will update this section when the PR is ready to be merged!

- [x] Implement login/logout util function
- [x] Clean up code
- [x] Setup Playwright's multiple sign up roles

#### Other Checks

- [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

#### User Facing Testing

Ensure Playwright E2E tests pass and ensure Puppeteer E2Es also pass, so that we can be sure the removed tests don't impact on anything.

##### Test the multiple sign-in roles feature

1. In your local Woo Blocks plugin, set breakpoints in the test file `tests/e2e/tests/checkout/checkout-block.block_theme.side_effects.spec.ts` at the lines `39` and `114`
2. Run the test file `tests/e2e/tests/checkout/checkout-block.block_theme.side_effects.spec.ts` in debug mode
3. The first time the code is paused, check the Playwright's browser. Ensure you are on the shop page
4. Open the `My Account` page in a new tab. Ensure the customer `Jane Smith` is logged in
5. The second time the code is paused, check the Playwright's browser. Ensure you are on the shop page
6. Open the `My Account` page in a new tab. Ensure the user is logged out

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested (ie: it makes changes to tests, coding standards, docblocks, etc.). -->
* [ ] Should be tested by the development team exclusively <!-- Check this box if this PR should be tested by the development team exclusively (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

tarhi-saad committed:
Convert checkout place order E2E tests to Playwright

tarhi-saad committed:
Add "gotoMyAccount" method

tarhi-saad committed:
Create login/logout utility functions

tarhi-saad committed:
Use the existing "customer" test data

tarhi-saad committed:
Complete the place order test cases

tarhi-saad committed:
Fix "My Account" page title

tarhi-saad committed:
Fix rebase

tarhi-saad committed:
Reset My account page title

tarhi-saad commented:
Hey @opr! My tests are failing, but I can't debug them locally. I can't run a single-file test using CLI or the VS Code extension (all 50 tests get executed in each instance). And for some reason, the tests are failing locally! I followed the suggestion from p1692138789517629-slack-C02UBB1EPEF, but without success!

opr reviewed (changes requested):
Hi Saad, thanks for the PR. @gigitux and I discussed about auth a little, and I think we should use [Multiple signed in roles](https://playwright.dev/docs/auth#multiple-signed-in-roles) to test for authenticated vs unauthenticated users.

Because of this I would recommend the following changes to this PR:

1. Remove the `login`, `logOut`, and `isLoggedIn` utils from the `FrontendUtils` class
2. Consider adding a `tests/auth.setup.ts` file as outlined in the playwright docs link above, log in as the user there and save the storage state.
3. In the test, create one context for each user type (logged in, logged out) and run the test for both users in the same test case.

Also, you may consider incorporating these tests in `checkout-block-shipping.block_theme.side_effects.spec.ts` instead of a new file, adding the cases to the existing test there. (Side note, we may want to rename this file now since it's not really focusing on shipping anymore).

I have been thinking about it more and I think it is actually a better idea to test as much as we can in one test, rather than setting up new ones for each thing we want to check.

Let me know what you think!

----

> I can't run a single-file test using CLI or the VS Code extension (all 50 tests get executed in each instance).

Go to `playwright.side-effects.config.ts` and remove
`dependencies: [ 'blockThemeConfiguration' ],`
from the `blockThemeWithGlobalSideEffects` config object.

Then the command I ran was: 

`npm run test:e2e:side-effects -- --headed tests/e2e/tests/checkout/checkout-block-place-order.block_theme.side_effects.spec.ts`

tarhi-saad committed:
Check for heading instead of the page title

tarhi-saad committed:
Check for heading in login/logout functions

tarhi-saad committed:
Fix all failing tests

tarhi-saad committed:
Add guest/customer/admin roles

tarhi-saad commented:
@opr, I created the `tests/auth.setup.ts` file that includes the following roles: `Guest`, `Customer`, and `Admin`. I get the error:

```Shell
Error: Error reading storage state from playwright/.auth/customer.json:
ENOENT: no such file or directory, open 'playwright/.auth/customer.json'
```

I believe the file `tests/auth.setup.ts` isn't recognized by the Playwright config. I tried adding it to the `playwright.config.ts`, but it didn't work. Do you know how to proceed from here?

> In the test, create one context for each user type (logged in, logged out) and run the test for both users in the same test case.

I believe the code will look like this:

```TS
test( 'Guest and logged in users can place order', async ( {
		pageObject,
		frontendUtils,
		page,
	} ) => {
                const customerContext = await browser.newContext({ storageState: 'playwright/.auth/customer.json' });
                const customerPage = await adminContext.newPage();

		await frontendUtils.emptyCart();
		await frontendUtils.goToShop();
		await frontendUtils.addToCart( SIMPLE_PHYSICAL_PRODUCT_NAME );
		await frontendUtils.goToCheckout();
		await expect(
			await pageObject.selectAndVerifyShippingOption(
				FREE_SHIPPING_NAME,
				FREE_SHIPPING_PRICE
			)
		).toBe( true );
		await pageObject.fillInCheckoutWithTestData();
		await pageObject.placeOrder();
		await expect(
			customerPage.getByText( 'Your order has been received.' )
		).toBeVisible();

               const guestContext = await browser.newContext({ storageState: 'playwright/.auth/guest.json' });
                const guestPage = await adminContext.newPage();

              etc...
	} );
```

I haven't tested it yet, but I was wondering how `frontendUtils` object will use the new `page` for each context (i.e., customPage and guestPage)?

opr commented:
> @opr, I created the tests/auth.setup.ts file that includes the following roles: Guest, Customer, and Admin. I get the error:

Hi @tarhi-saad where does this error appear? What steps can I take to reproduce it?

> I haven't tested it yet, but I was wondering how frontendUtils object will use the new page for each context (i.e., customPage and guestPage)?

Apologies, I think I wrote something incorrectly in my review, I spoke more to Luigi and I think we should use one test per user-role. We should therefore write the test like this (from playwright docs)

![image (4)](https://github.com/woocommerce/woocommerce-blocks/assets/5656702/ddab439c-108a-4b0e-bd62-3bcf88a03ed2)


To ensure the `auth.setup.js` file is executed, I added it as a project in `playwright.config.js` and `playwright.side-effects.config.js` like so:

```js
{
	name: 'authSetup',
	testDir: '.',
	testMatch: /auth.setup.ts/,
},
```

I then made the following projects depend on it: `blockThemeConfiguration` (in both files), and `classicThemeConfiguration` in the `playwright.config.js` file.

Example config:

```
{
	name: 'blockThemeConfiguration',
	testDir: '.',
	testMatch: /block-theme.setup.ts/,
	dependencies: [ 'authSetup' ],
},
```


Also, in `auth.setup.js` I had to ensure the browser was logged out in each case before logging in, similar to how you did it in the `authenticate as guest` step, I also added this to both admin/customer steps too, just to ensure we're logging out and then back in with the correct account.

Finally, the `test.use( { storageState: 'tests/e2e/.auth/customer.json' } );` must be invoked at the `test.describe` level, meaning we should describe once for the logged in user, and once for the logged out user!

Hope this is useful 😎

tarhi-saad committed:
Update the auth setup

tarhi-saad committed:
Register the auth setup within Playwright

tarhi-saad committed:
Update testing cases

tarhi-saad committed:
Add generated auth files to .gitignore

opr commented on tests/e2e/auth.setup.ts:
```diff
@@ -0,0 +1,121 @@
+			.fill( admin.username );
+		await page.getByLabel( 'Password' ).fill( admin.password );
+		await page.getByRole( 'button', { name: 'Log in' } ).click();
+		// Wait until the page receives the cookies.
+		//
+		// Sometimes login flow sets cookies in the process of several redirects.
+		// Wait for the final URL to ensure that the cookies are actually set.
```
```suggestion
		// Sometimes login flow sets cookies in the process of several redirects.
		// Wait for the final URL to ensure that the cookies are actually set.
```

Let's tidy up the comments a little bit.

opr commented on tests/e2e/auth.setup.ts:
```diff
@@ -0,0 +1,121 @@
+			.getByRole( 'link', { name: 'Log out' } )
+	).toBeVisible();
+
+	// End of authentication steps.
+
```
```suggestion
```

Remove unnecessary comment.

opr commented on tests/e2e/auth.setup.ts:
```diff
@@ -0,0 +1,121 @@
+			.fill( customer.username );
+		await page.getByLabel( 'Password' ).fill( customer.password );
+		await page.getByRole( 'button', { name: 'Log in' } ).click();
+		// Wait until the page receives the cookies.
+		//
+		// Sometimes login flow sets cookies in the process of several redirects.
+		// Wait for the final URL to ensure that the cookies are actually set.
```
```suggestion
		// Sometimes login flow sets cookies in the process of several redirects.
		// Wait for the final URL to ensure that the cookies are actually set.
```

Same as above

opr commented on tests/e2e/auth.setup.ts:
```diff
@@ -0,0 +1,121 @@
+	).toBeVisible();
+
+	// End of authentication steps.
+
```
```suggestion
```
Same as above

opr commented on tests/e2e/auth.setup.ts:
```diff
@@ -0,0 +1,121 @@
+		// Wait until the page receives the cookies.
+		//
+		// Sometimes login flow sets cookies in the process of several redirects.
+		// Wait for the final URL to ensure that the cookies are actually set.
```
```suggestion
		// Sometimes login flow sets cookies in the process of several redirects.
		// Wait for the final URL to ensure that the cookies are actually set.
```

Same as above.

opr commented on tests/e2e/auth.setup.ts:
```diff
@@ -0,0 +1,121 @@
+		await page.waitForURL( '/my-account/' );
+	}
+
+	// Alternatively, you can wait until the page reaches a state where all cookies are set.
```
Not sure what value this comment adds. Let's remove or revise it to actually explain what's happening.

opr commented on tests/e2e/auth.setup.ts:
```diff
@@ -0,0 +1,121 @@
+	await page.context().storageState( { path: adminFile } );
+} );
+
+const customerFile = 'playwright/.auth/customer.json';
```
Let's store these file paths in a constants file, also let's make them go into `tests/e2e/.auth/` because otherwise we'll have a new `playwright` directory created.

Let's also add `tests/e2e/.auth` to `.gitignore`

opr commented on tests/e2e/auth.setup.ts:
```diff
@@ -0,0 +1,121 @@
+const customerFile = 'playwright/.auth/customer.json';
+
+setup( 'authenticate as customer', async ( { page } ) => {
+	// Authentication steps.
```
Comment adds no value, really. Let's remove (same for other steps)

opr commented on tests/e2e/playwright.side-effects.config.ts:
```diff
@@ -21,11 +21,17 @@ export default defineConfig( {
 			name: 'blockThemeWithGlobalSideEffects',
 			testMatch: /.*.block_theme.side_effects.spec.ts/,
-			dependencies: [ 'blockThemeConfiguration' ],
+			dependencies: [ 'blockThemeConfiguration', 'authSetup' ],
```
Don't think this is necessary because `blockThemeConfiguration` depends on `authSetup`

opr commented on tests/e2e/tests/checkout/checkout-block-place-order.block_theme.side_effects.spec.ts:
```diff
@@ -0,0 +1,76 @@
+} );
+
+test.describe( 'Shopper → Checkout block → Place Order → customer', () => {
+	test.use( { storageState: 'playwright/.auth/customer.json' } );
```
If we update the path then be sure to update it here too, if we have a constants file we can also import the path from there.

opr commented on tests/e2e/tests/checkout/checkout-block-place-order.block_theme.side_effects.spec.ts:
```diff
@@ -0,0 +1,76 @@
+test.describe( 'Shopper → Checkout block → Place Order → customer', () => {
+	test.use( { storageState: 'playwright/.auth/customer.json' } );
+
+	test( 'Guest user can place order', async ( {
```
Do you think this test case is covered by https://github.com/woocommerce/woocommerce-blocks/blob/trunk/tests/e2e/tests/checkout/checkout-block-shipping.block_theme.side_effects.spec.ts#L74-L81 instead? If we execute `test.use( { storageState: 'playwright/.auth/customer.json' } );` in that file, we're already testing if we can place the order in that file so probably don't need a separate test for it?

opr replied:
Think you got the name of this test mixed up with the other, this is for logged in users, not guests (based on the storage state we're loading)

opr commented on tests/e2e/tests/checkout/checkout-block-place-order.block_theme.side_effects.spec.ts:
```diff
@@ -0,0 +1,76 @@
+test.describe( 'Shopper → Checkout block → Place Order → guest', () => {
+	test.use( { storageState: 'playwright/.auth/guest.json' } );
+
+	test( 'Logged in user can place an order', async ( {
```
Think the name should be a logged out user.

I think if we remove the above test, and incorporate it into https://github.com/woocommerce/woocommerce-blocks/blob/trunk/tests/e2e/tests/checkout/checkout-block-shipping.block_theme.side_effects.spec.ts#L74-L81 we could move this test into that same file (and rename the file, as it's no longer focused on shipping).

In my opinion, because these tests are so closely linked it makes sense to combine the files.

opr commented on tests/e2e/tests/checkout/constants.ts:
```diff
@@ -0,0 +1,4 @@
+export const SIMPLE_VIRTUAL_PRODUCT_NAME = 'Album';
+export const SIMPLE_PHYSICAL_PRODUCT_NAME = 'Beanie';
+export const FREE_SHIPPING_NAME = 'Free shipping';
+export const FREE_SHIPPING_PRICE = '$0.00';
```
If we're adding these constants can we update https://github.com/woocommerce/woocommerce-blocks/blob/trunk/tests/e2e/tests/checkout/checkout-block-shipping.block_theme.side_effects.spec.ts#L21-L24 so it uses them too?

opr commented on tests/e2e/tests/checkout/checkout-block-place-order.block_theme.side_effects.spec.ts:
```diff
@@ -0,0 +1,76 @@
+		frontendUtils,
+		page,
+	} ) => {
+		// await frontendUtils.login();
```
Commented code
```suggestion
```

opr commented on tests/e2e/utils/frontend/frontend-utils.page.ts:
```diff
@@ -121,4 +127,59 @@ export class FrontendUtils {
+		return ! isLoginFormVisible;
+	}
+
+	async login() {
```
Are we using this function anymore? I don't think we are.

opr commented on tests/e2e/utils/frontend/frontend-utils.page.ts:
```diff
@@ -121,4 +127,59 @@ export class FrontendUtils {
+		await this.page.waitForLoadState( 'networkidle' );
+	}
+
+	async logout() {
```
Are we using this function anymore? I don't think we are.

opr commented on tests/e2e/utils/frontend/frontend-utils.page.ts:
```diff
@@ -3,6 +3,12 @@
+/**
+ * Internal dependencies
+ */
+import { customer } from '../../test-data/data/data';
```
Won't need this if we remove `login` and `logout`

opr reviewed (changes requested):
Hey @tarhi-saad thanks for working on this, starting to look good, I think the auth steps are really valuable. I left some inline comments that should be addressed. Thanks!

tarhi-saad committed:
Tidy up comments

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

tarhi-saad committed:
Remove unnecessary comment

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

tarhi-saad committed:
Update comments

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

tarhi-saad committed:
Remove unnecessary comment

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

tarhi-saad committed:
Update comment for Guest case

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

tarhi-saad committed:
Remove confusing comment

tarhi-saad committed:
Remove another unnecessary comment

tarhi-saad committed:
Remove unnecessary Playwright project dependency

tarhi-saad committed:
Tidy up the file structure and constants

tarhi-saad committed:
Fix mixed up test descriptions

tarhi-saad committed:
Remove commented code

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

tarhi-saad committed:
Remove unnecessary function from frontend-utils

tarhi-saad committed:
Refactor testing cases

tarhi-saad committed:
Rename testing file

tarhi-saad committed:
Delete unused testing file

tarhi-saad commented:
@opr, I've incorporated all the feedback! However, the shipping test is failing. It seems that from the beginning, the authentication setup hasn't been functioning properly. We've been placing orders without authentication consistently (even when using `customer`). I'll continue to investigate to identify a solution!

opr committed:
Ensure we're logged out before trying to log in as a user

opr commented on tests/e2e/utils/constants.ts:
```diff
@@ -10,3 +10,8 @@ export const STORAGE_STATE_PATH = path.join(
 	'artifacts/storage-states/admin.json'
 );
+
+// User roles file paths
+export const adminFile = '.auth/admin.json';
+export const customerFile = '.auth/customer.json';
+export const guestFile = '.auth/guest.json';
```
We are saving these at the project root, but the `.gitignore` only ignores things in `e2e`. We should update these paths.

tarhi-saad replied:
@opr, when I run the tests locally, these files are created in `tests/e2e/.auth/`, not on the project root.

opr committed:
Log out before each authentication setup step

opr commented:
> @opr, I've incorporated all the feedback! However, the shipping test is failing. It seems that from the beginning, the authentication setup hasn't been functioning properly. We've been placing orders without authentication consistently (even when using `customer`). I'll continue to investigate to identify a solution!

Hi @tarhi-saad - during the auth setup, we were checking if we're logged out, it looks like the same context is used between setup steps, so the admin account is logged in. I reversed the logic, so now instead of checking if we're logged _out_,  we will check if we're logged in (i.e., is the logout link visible) and we will log out before each test.

tarhi-saad commented:
This is awesome, @opr! I tested it locally, and it worked fine! But, since there are still some failing tests because of this PR: https://github.com/woocommerce/woocommerce-blocks/pull/10611, I linked the fix https://github.com/woocommerce/woocommerce-blocks/pull/10695 as a blocker.

opr committed:
Ensure tests requiring admin are logged in

tarhi-saad commented:
We are putting this PR into draft for now! We will resume working on it when @gigitux is back! We need to figure out how our auth setup steps can be used by the work Luigi did to revert templates. Before proceeding, we need to understand that part fully (see Slack p1692889180189739/1692886577.310339-slack-C8X6Q7XQU).

Slack bug discussion: p1692878370587469/1692714875.315809-slack-C8X6Q7XQU

ralucaStan commented:
@tarhi-saad is it still worth keeping this PR? Was there a decision?

tarhi-saad commented:
@ralucaStan, I couldn't return to this PR during the last cooldown. Implementing the "Multiple signed-in roles" feature is crucial since we can't write reliable tests with multiple roles right now. Should we add this task to the current cycle?

tarhi-saad committed:
Log in as admin during block theme setup

tarhi-saad committed:
Fix Playwright strict mode violation

tarhi-saad committed:
Run Multiple sign-in roles to the global-setup phase

In this step of the Playwright's setup, we can add the multiple sign-in
roles and keeping the admin logged by default.
This fixes the issue of failing tests `logged out` error.

tarhi-saad committed:
Remove unnecessary login as admin

tarhi-saad committed:
Remove "auth.setup" dependency since the logic lives in the global setup instead

tarhi-saad committed:
Remove unnecessary login as admin from test files

The admin profile is set by default

tarhi-saad commented:
@opr, the multiple sign-in roles are now set at the global setup phase. Adding it through a dependency caused issues (i.e., log-out errors in multiple E2E tests and config files). I tried using a new context browser but without success. Fortunately, we can avoid interfering with the global login setup by running our multiple sign-in roles early on in the global setup!

gigitux commented:
@tarhi-saad Great work! 

Before merging this PR, I think that it makes sense to add some E2E tests to ensure that the approach works (and cath up future regression).

I would create an `auth.spec.block_theme.ts` test that checks the user is the right one. We could use the menubar at the top of the page:
![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/12e92123-1977-4eb6-b631-1efded03f5ac)

- if it is not present, the user is not logged.
- if it is present, check the username.

Also, by default, the logged user is the admin, right?

opr reviewed (commented):
With this approach is it possible to change which user we're using mid-test, or to have a few different page objects with different signed-in states?

One example is: if the admin makes a change to the site settings as part of a test, how can we test that the logged-out user sees the updated changes? 

If you could check that this is possible that would be good to know, and some instructions on how to achieve that would also be great!

tarhi-saad commented:
Hi, @opr, @gigitux! 👋 Thank you for the reviews! 🙌 Can we tackle the requested changes in a separate ticket? I would like to merge this PR before tomorrow's EOD due to the migration to the mono repo (p1701870709042359-slack-C02UBB1EPEF). I couldn't work on these changes because I'm on Porter's rotation this week! We also can start using the multiple sign-in roles feature in our E2E migration project.
If you are ok with this, I suggest the following testing instructions for this PR:

> ##### Test the multiple sign-in roles feature
>
> 1. In your local Woo Blocks plugin, set breakpoints in the test file `tests/e2e/tests/checkout/checkout-block.block_theme.side_effects.spec.ts` at the lines `39` and `114`
> 2. Run the test file `tests/e2e/tests/checkout/checkout-block.block_theme.side_effects.spec.ts` in debug mode
> 3. The first time the code is paused, check the Playwright's browser. Ensure you are on the shop page
> 4. Open the `My Account` page in a new tab. Ensure the customer `Jane Smith` is logged in
> 5. The second time the code is paused, check the Playwright's browser. Ensure you are on the shop page
> 6. Open the `My Account` page in a new tab. Ensure the user is logged out

opr reviewed (approved):
Ok the testing steps you proposed are working fine!
